presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -582,9 +582,11 @@ private PlanWithProperties planTableScan(TableScanNode node, Expression predicat
 
             Map<ColumnHandle, Symbol> assignments = ImmutableBiMap.copyOf(node.getAssignments()).inverse();
 
+            // Simplify the tuple domain to avoid creating an expression with too many nodes that's
+            // expensive to evaluate in the call to shouldPrune below.
             Expression constraint = combineConjuncts(
                     deterministicPredicate,
-                    DomainTranslator.toPredicate(node.getCurrentConstraint().transform(assignments::get)));
+                    DomainTranslator.toPredicate(newDomain.simplify().transform(assignments::get)));
```

Comment:
I am worried that we lose some important information here. We may win planning performance, but I am afraid about execution.

Reply:
It shouldn't. The full tuple-domain based constraint is passed to getTableLayouts(). This is just what gets added to the opaque expression.

Reply:
You are passing that to `LayoutConstraintEvaluator`, doesn't it make less effective (selective)?

Reply:
It is, but that's why we pass the full tuple domain as well. Connectors are, generally, supposed to use the tuple domain first and for anything they can't tell for sure, fall back to the opaque expression.

In the future (once we have a new IR and are able to pass the whole expression to a connector), this will be all moot, anyway.